### DISCUSSION TITLE
Uses and configuration choices for CF GitHub Discussions

### DISCUSSION THREAD

**Discussion by JonathanGregory** [cf-convention/discuss#289] [Comments and ideas for changing CF]
## Welcome to CF Discussions!

This is a forum for community discussion of matters relating to CF. I have set it up, with help from @davidhassell, as [decided in December](https://cfconventions.org/Governance/GovPanel/2023-12-04-meeting.html) by the CF governance panel, following earlier discussion, especially at the last annual meeting. As you can see, a GitHub Discussion doesn't look _very_ different from an issue, but they offer various other facilities. Quoting from the [documentation](https://resources.github.com/devops/process/planning/discussions):

> A common question among developers new to GitHub Discussions goes something like this: When should I use Discussions and when should I use Issues? **The short answer:** Discussions are great for questions and ideas that require team communication to make a decision, while Issues are defined pieces of work.

Now we've enabled it, **please try using Discussions for all new announcements, comments, ideas and Q&A**, instead of issues in repositories, so we can gather experience with it. We hope no-one will mind trying. When we've tried for a while, we can decide (by consensus, as usual) how to proceed.

When you open a new discussion, you have to choose a category for it. There are currently just three categories:

* Announcements. The default GitHub Discussions setup is that only "maintainers" can make announcements, but we have changed it so that anyone can make one. The distinguishing feature of announcements is that they _aren't_ expected to begin a discussion.

* Questions and answers. This category is for asking questions about how to use CF **as it is** e.g. the meaning of a convention, the appropriate standard name for a quantity, why something isn't working as expected, where to find some information, how to make a proposal for change. A question invites answers. In the default GitHub Discussions setup, answers can be voted for and ordered by votes, like you see for instance in `stackexchange`. We have disabled this facility, because of our impression that questions to CF generally don't receive many answers, so there's no need to vote on them, and also that the answer is sometimes arrived at by discussion involving several people.

* Comments and ideas. This category is for discussions about how CF could be improved, what's wrong with it, what its purposes should be, whether it could be managed better, etc.  **If you have a definite proposal for change, it's fine to open a new issue straight away, rather than starting a discussion, just as before** i.e. in the [`discuss` repo](https://github.com/cf-convention/discuss/issues) for standard names and other vocabulary, the [`cf-conventions` repo](https://github.com/cf-convention/cf-conventions/issues) for conventions, and the [`cf-convention.github.io`](https://github.com/cf-convention/cf-convention.github.io/issues) repo for website and governance.

The default GitHub Discussions setup has a "poll" category as well. We have disabled this because CF makes decisions by consensus, not by voting, so we've not had a need for polls up to now.

The category can be changed for an existing discussion. It seems likely that a Q&A discussion (about how CF is now) might sometimes get changed into an into a comment or an idea (for how CF should be changed). We'll see. A GitHub discussion can be converted into a GitHub issue. It is likely that this will happen if a discussion identifies a change that could be proposed, after having begun without a definite idea for change. **Decisions and discussions about proposed changes will still be made in issues** according to our existing rules.

As you see, the categories are functional, rather relating to the subject matter. It is useful to apply one or more of the following five labels to indicate the area of the discussion, in order to guide people who are searching the discussions for something relevant (you can filter by label): `conventions`, `vocabulary` (standard names etc.), `website`, `governance` or `GitHub`. Apply more than one label if it's unclear which is applicable. You will see that there are several other available labels. That is because these Discussions are hosted in the existing `discuss` repo, and those are the labels used for issues in this repo. When we've got used to Discussions, we can tidy this up.

It would be useful if we could put explanatory text about labels in the entry box for a new discussion. You can create a [form for new GitHub discussions](https://docs.github.com/en/discussions/managing-discussions-for-your-community/creating-discussion-category-forms) in YaML, but I can't find a way to produce a text entry box like the one you get by default, with buttons for creating markdown, and most importantly with a preview capability. The `textarea` field is for simple text entry. With these forms, it would also be possible to add labels automatically, but I think we would need a different category for every possible combination of labels. That would be extremely clumsy. However, perhaps there are better ways to do these things.

I guess that, unlike issues, there won't be a need to close discussions. It doesn't really matter if the list of open discussions becomes long, whereas with issues this _is_ a problem, because the issues are the changes we are actively considering. If there are too many of them open in the list, you can't easily get an overview. But we'll see.

I chose the "Comments and ideas" category for this item that you are reading, rather than categorising it as an announcement, because our setup is all open for discussion here. Please say what you think about the above.

Best wishes

Jonathan

**Comment by davidhassell** [comment]
Thanks, Jonathan.

This is test reply!

**Comment by ChrisBarker-NOAA** [comment]
> The default GitHub Discussions setup has a "poll" category as well. We have disabled this because CF makes decisions by consensus, not by voting, so we've not had a need for polls up to now.

Polls can be useful for gathering opinions on the way to consensus -- but it makes sense to  wait for a use case before turning it on.

(really put this in as a way to test using the discussion as anything else ...)

**Comment by JonathanGregory** [comment]
Dear all

The CF information management team met this afternoon, chaired by @ethanrd, and discussed Discussions.

* The Discussions are hosted in the `discuss` repository, where up to now we have created issues for discussions, announcements, questions and vocabulary (standard names etc.). For the moment, we're encouraging discussions, announcements and questions to happen in Discussions instead, but we continue with issues in the `discuss` repo for vocabulary. The templates for new issues for questions should be modified to ask the author to use Discussions instead.

* Later, we will propose to move the existing and new vocabulary issues to a new `vocabulary` repo. This change will give CF community members a new choice, because only those who watch the new repo will be notified of vocabulary issues. We will also propose to convert the other existing `discuss` issues into Discussions. We discussed whether the standard names files should be moved from the website repo to the new `vocabulary` repo. It is not necessary to do this. Possible reasons to do it are that it would be neater, it would allow DOIs to be minted specifically identifying the vocabulary, and it would alleviate shortage of space on GitHub pages for the website.

* Through experiments by Fran @feggleton, Andrew @DocOtak and @davidhassell, we discovered that if a discussion wants to spawn an issue to change one of the repos (conventions etc.), the best way to do it is by selecting "Reference in new issue" from the &#x2026; menu at the top right of the first or any subsequent contribution to the discussion. With this method you can specify _any repository_ for the new issue, and optionally copy the text from the selected post. By contrast, if you select &odot; "Create issue from discussion" in the menu on the right of the discussion, it can only be created in the `discuss` repo. When creating an issue, a link to it and some comment should be made in the discussion giving rise to it.

* The website home page and discussions page need to be updated to mention the use of Discussions. We will also need a page of guidance about how to use Discussions, which we can base on this discussion, when we've had some more experience.

* I stated in the first post above, and in the meeting, that you couldn't produce a text entry box allowing markdown and previews. That was wrong. I was misled because it looks different when it first appears in a new discussion. It appears to be a plain text box, but once you write in it, it changes to the more elaborate kind of box supporting markdown. Therefore I've installed a template for new discussions, which currently looks like this:

![new discussion template](https://github.com/cf-convention/discuss/assets/8126248/4045e10a-2bd7-40e5-bc3a-fecb9387e38a)

* In order to experiment with Discussions features, I created a [test GitHub organisation](https://github.com/cf-convention-test). I will invite members of the CF information management team to join that organisation, so that you can experiment for yourself. If any other CF member would like to join in, please let me know. I think that's a better place to experiment than this discussion is, because we won't disturb anyone else, bearing in mind that everyone who's watching CF is notified of contributions to our discussion here!

Best wishes

Jonathan

**Reply by DocOtak** [reply]
@japamment @feggleton Related to the standard name repository.

As discussed at the meeting this today, the standard names data is large. GitHub has a [limit of 1GB](https://docs.github.com/en/pages/getting-started-with-github-pages/about-github-pages#usage-limits) for the **published size** of the website, distinct from the **repository size** which is actually quite reasonable for the CF website. The last build of the cf conventions website was 858MB, with the current size of each standard name publication, I suspec there is space remaining for fewer than 10 standard name table publications before we hit this limit and will be unable to build/update the website.

As a possible solution, we discussed using a single git repository and utilizing tags for the versioned tables. I have prepared an example of how this could work: https://github.com/DocOtak/cf_standard_names

If you browse the code, you will notice that only the "current" version is browsable, the checked out version of this will never be larger than the current size of the most recent name table (currently aprox. 18MB).

The different versions have all been tagged: https://github.com/DocOtak/cf_standard_names/tags though I was very sparse on the tag message.

If for example if you wanted to see the name table in version 78, you can browse the code: https://github.com/DocOtak/cf_standard_names/tree/78
You can also link directly to the nametable xml: https://github.com/DocOtak/cf_standard_names/raw/78/cf_standard_names/src/cf-standard-name-table.xml

The different versions were loaded into the repository sequentially, git itself uses a delta compression for storage. The compressed version of the entire history of the standard names is only 2.2MB.

A potential downside of this method is that only the current version of the name table can be hosted directly in github pages, so only the "current" version would have the html interface on the website, though all the previous ones continue to exist within the git history.

**Reply by JonathanGregory** [reply]
Thanks for this useful comment, Andrew @DocOtak. In order to keep the present discussion focussed on "Discussions", I have started a [new discussion](https://github.com/orgs/cf-convention/discussions/296) on how to keep the versions of the standard name table, and I have copied your comment above to it. I suggest carrying on with that topic there, instead of here.

**Comment by larsbarring** [comment]
Just to test: I have just deleted my previous "just checking in ...." comment.

I think that @JonathanGregory's top post (or a copy/pasted version thereof) would be excellent as a _pinned_ __issue__ in the `discuss` repo to invite people over here to discussions.

**Comment by sadielbartholomew** [comment]
Forgive me if this has already been covered, but I don't think I've seen it mentioned...

### Use of upvoting and selected emojis on posts

One aspect of the topic I think would be useful to clarify, formally, is the meaning of upvote and selected emojis in CF GitHub Discussions, because unlike with GitHub Issues, we have this 'upvote' option and I think we can put it to good use to help to provide a _quick_ way for people to indicate their thoughts about a given comment, since not everyone has time to make a post but might want to acknowledge that they have read a post or have a certain sentiment.

I think it could be useful to outline an intended meaning for the upvote, versus the 'thumbs up' and 'thumbs down' emoji, at least (but potentially more of the set of eight provided for any comment can be put to use). Namely, given that [GitHub describes the upvote option as](https://docs.github.com/en/discussions/collaborating-with-your-community-using-discussions/participating-in-a-discussion#upvoting-a-discussion) as an indicator that someone thinks a comment/topic is important:

> upvote discussions to give more visibility to the topics that matter to you, and sort discussions to see which have been upvoted the most

the 'thumbs up (down)' can be seen as a way to indicate 'I (dis)agree with this comment/view' which is somewhat similar, but also quite different. For example, someone could think someone made a point that merits further discussion, but that they disagree with it, say. And 'thumbs down' use can help us to identify CF Conventions change/update proposals where someone isn't happy with something.

So, before people start to use Discussions for CF in earnest, I think it would be useful for us to outline clearly a distinction that we'd advocate people use with the upvote and the thumbs up/down emoji.

#### My suggestions for sentiments to encode in this way

On those lines, I for one would really like a quick way to say any of the following, and propose these actions for that, as a starting point:
* I have seen and read this comment (I suggest 'eyes' :eyes: ) 
* I think this is an important point that should be highlighted / needs visibility (suggest the upvote)
* I agree with this (suggest 'thumbs up' :+1: )
* I disagree with this (suggest 'thumbs down' :-1: )

**Reply by sadielbartholomew** [reply]
Pertinently, when you comment GitHub seems to set it so that you have upvoted your own comment, so you start on a '+1' upvote count unless you remove it. I guess that makes sense under the scheme above, since obviously any comment someone raises is one they think is important :smile:

**Reply by sethmcg** [reply]
I like your suggestions, and agree with them.  In particular, distinguishing between "I think this is important" (upvote) and "I agree" (thumbs up) is valuable, because then you can both upvote and thumbs-down a comment to indicate that while you think it's an important point, you disagree with the conclusion.
